Turtle serializing creates invalid PNames

Consider following triple from DBPedia serialized as .ttl:

test.ttl
```
@prefix ns1: <http://dbpedia.org/property/> .
<http://dbpedia.org/resource/Maine_Mendoza> ns2:spouse(s)_ "Nate Winchester 2013-Present"@en .
```
It is not possible to parse this triple due to the brackets in ns2:spouses(s)_ in a .ttl file. However, there is no such problem for an .nt file.

Code to parse the test.ttl file with the triple inside.
```
from rdflib import Graph
g = Graph()
g.parse(test.ttl)

```
Exception:
```
Traceback (most recent call last):
  File "/home/fkovacev/PhD/RDFStarVersioningEval/OSTRICH/BEAR/scripts/build_tb_rdf_star_datasets.py", line 204, in <module>
    construct_tb_star_ds(source_ic0=data_dir + "/alldata.IC.nt/000001.nt",
  File "/home/fkovacev/PhD/RDFStarVersioningEval/OSTRICH/BEAR/scripts/build_tb_rdf_star_datasets.py", line 140, in construct_tb_star_ds
    cs_add.parse(source_cs + "/" + t[1])
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/graph.py", line 1261, in parse
    raise se
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/graph.py", line 1252, in parse
    parser.parse(source, self, **args)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 1921, in parse
    p.loadStream(stream)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 438, in loadStream
    return self.loadBuf(stream.read())  # Not ideal
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 444, in loadBuf
    self.feed(buf)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 470, in feed
    i = self.directiveOrStatement(s, j)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 490, in directiveOrStatement
    j = self.statement(argstr, i)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 740, in statement
    j = self.property_list(argstr, i, r[0])
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 1088, in property_list
    i = self.objectList(argstr, j, objs)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 1132, in objectList
    i = self.object(argstr, i, res)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 1418, in object
    j = self.subject(argstr, i, res)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 747, in subject
    return self.item(argstr, i, res)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 839, in item
    return self.path(argstr, i, res)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 846, in path
    j = self.nodeOrLiteral(argstr, i, res)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 1446, in nodeOrLiteral
    j = self.node(argstr, i, res)
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 1025, in node
    self.BadSyntax(argstr, i, "expected item in list or ')'")
  File "/home/fkovacev/anaconda3/envs/baseDSEnv/lib/python3.8/site-packages/rdflib/plugins/parsers/notation3.py", line 1653, in BadSyntax
    raise BadSyntax(self._thisDoc, self.lines, argstr, i, msg)
rdflib.plugins.parsers.notation3.BadSyntax: at line 3 of <>:
Bad syntax (expected item in list or ')') at ^ in:
"...b'> .\n\n<http://dbpedia.org/resource/Maine_Mendoza> ns1:spouse('^b's)_ "Nate Winchester 2013-Present"@en .\n\n\n'"

```

I have been looking at this a bit, the only cases in which invalid PNames are returned are:

```
FAILED test/test_turtle_quoting.py::test_serialize_correctness[turtle-(] - rdflib.plugins.parsers.notation3.BadSyntax: <no detail available>
FAILED test/test_turtle_quoting.py::test_serialize_correctness[turtle-)] - rdflib.plugins.parsers.notation3.BadSyntax: <no detail available>
FAILED test/test_turtle_quoting.py::test_serialize_correctness[turtle-%] - rdflib.plugins.parsers.notation3.BadSyntax: <no detail available>
```

In all these cases there is no problem:

```
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-A]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-_]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-~]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-.]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle--]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-!]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-$]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-&]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-']
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-*]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-+]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-,]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-;]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-=]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-/]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-?]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-#]
PASSED test/test_turtle_quoting.py::test_serialize_correctness[turtle-@]
```

Test code: https://github.com/aucampia/rdflib/blob/092f228752a89e2c1eecac376a7c148e64a83241/test/test_turtle_quoting.py#L155-L176

And the reason it fails is actually because of is_ncname in rdflib.namespace:

https://github.com/aucampia/rdflib/blob/092f228752a89e2c1eecac376a7c148e64a83241/rdflib/namespace/__init__.py#L643-L669

Basically, if "%", "(" and ")" is removed from there all turtle test pass, of course that is not a fix for this, but rather this code is XML specific and should probably not run at all for Turtle based syntax.

@niklasl 's suggestion will fix it though, but will have to dig a bit into this at some point to try and rationalize what is going on here.

Draft fix in https://github.com/RDFLib/rdflib/pull/1678 - however this causes one regression, in that it breaks percent encoded stuff:

```
$ .venv/bin/pytest  'test/test_issue801.py::TestIssue801::test_issue_801'
============================================================================ test session starts ============================================================================
platform linux -- Python 3.8.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /home/iwana/sw/d/github.com/iafork/rdflib, configfile: tox.ini
plugins: subtests-0.5.0, cov-3.0.0
collected 1 item                                                                                                                                                            

test/test_issue801.py F                                                                                                                                               [100%]

================================================================================= FAILURES ==================================================================================
________________________________________________________________________ TestIssue801.test_issue_801 ________________________________________________________________________

self = <test.test_issue801.TestIssue801 testMethod=test_issue_801>

    def test_issue_801(self):
        g = Graph()
        example = Namespace("http://example.org/")
        g.bind("", example)
        node = BNode()
        g.add((node, example["first%20name"], Literal("John")))
>       self.assertEqual(
            g.serialize(format="turtle").split("\n")[-3], '[] :first%20name "John" .'
        )
E       AssertionError: '[] :first\\%20name "John" .' != '[] :first%20name "John" .'
E       - [] :first\%20name "John" .
E       ?          -
E       + [] :first%20name "John" .

test/test_issue801.py:15: AssertionError
========================================================================== short test summary info ==========================================================================
FAILED test/test_issue801.py::TestIssue801::test_issue_801 - AssertionError: '[] :first\\%20name "John" .' != '[] :first%20name "John" .'
============================================================================= 1 failed in 0.18s =============================================================================
```

Possibly we should just go very conservative and only deal with `(` and `)` as these are actually the main problems.

So something like what @niklasl suggested but with:

```python
PNAME_LOCAL_ESCAPES = re.compile(r"([()])")
```



Potentially also with a different name.

I think there may be better places to fix this though, but will look at it more tomorrow. I think a targetted fix for only the brackets may make sense though, and then we can unravel the rest separately.

--- a/rdflib/plugins/serializers/turtle.py
+++ b/rdflib/plugins/serializers/turtle.py
@@ -268,6 +268,7 @@ class TurtleSerializer(RecursiveSerializer):
         if isinstance(p, BNode):  # hmm - when is P ever a bnode?
             self._references[p] += 1
 
+    # TODO: Rename to get_pname
     def getQName(self, uri, gen_prefix=True):
         if not isinstance(uri, URIRef):
             return None
@@ -288,6 +289,8 @@ class TurtleSerializer(RecursiveSerializer):
                 return None
 
         prefix, namespace, local = parts
+
+        local = local.replace(r"(", r"\(").replace(r")", r"\)")
 
         # QName cannot end with .
         if local.endswith("."):

--- a/test/nt/rdflibtest-pnamebrackets.nt
+++ b/test/nt/rdflibtest-pnamebrackets.nt
@@ -1,0 +1,1 @@
+<http://example.com/instance/John_Doe> <http://example.com/property/name(s)> "John Doe"@en.

--- a/test/test_roundtrip.py
+++ b/test/test_roundtrip.py
@@ -138,6 +138,10 @@ XFAILS = {
         reason="missing escaping: PCDATA invalid Char value 12 and 8",
         raises=SAXParseException,
     ),
+    ("xml", "rdflibtest-pnamebrackets.nt"): pytest.mark.xfail(
+        reason="results in invalid xml element name: <ns1:name(s)/>",
+        raises=SAXParseException,
+    ),
 }
 
 # This is for files which can only be represented properly in one format

--- a/test/test_serializer_turtle.py
+++ b/test/test_serializer_turtle.py
@@ -112,4 +112,4 @@ def test_turtle_namespace():
     assert "RO_has_phenotype:" in output
     assert "GENO:0000385" in output
     assert "SERIAL:0167-6423" in output
-    assert "EX:name_with_(parenthesis)" in output
+    assert r"EX:name_with_\(parenthesis\)" in output

--- a/test/test_turtle_quoting.py
+++ b/test/test_turtle_quoting.py
@@ -1,16 +1,20 @@
 """
-This module is intended for tests related to quoting/escaping and
-unquoting/unescaping in various formats that are related to turtle, such as
-ntriples, nquads, trig and n3.
+This module is intended for tests related to unquoting/unescaping in various
+formats that are related to turtle, such as ntriples, nquads, trig and n3.
 """
 
+import itertools
+import logging
 from typing import Callable, Dict, Iterable, List, Tuple
 import pytest
-from rdflib.graph import ConjunctiveGraph
+from rdflib.graph import ConjunctiveGraph, Graph
 
 from rdflib.plugins.parsers import ntriples
-from rdflib.term import Literal
+from rdflib.term import Literal, URIRef
+from rdflib import Namespace
+from .testutils import GraphHelper
 
+# https://www.w3.org/TR/turtle/#string
 string_escape_map = {
     "t": "\t",
     "b": "\b",
@@ -22,8 +26,10 @@ string_escape_map = {
     "\\": "\\",
 }
 
+import re
 
-def make_correctness_pairs() -> List[Tuple[str, str]]:
+
+def make_unquote_correctness_pairs() -> List[Tuple[str, str]]:
     """
     Creates pairs of quoted and unquoted strings.
     """
@@ -59,7 +65,7 @@ def make_correctness_pairs() -> List[Tuple[str, str]]:
     return result
 
 
-CORRECTNESS_PAIRS = make_correctness_pairs()
+UNQUOTE_CORRECTNESS_PAIRS = make_unquote_correctness_pairs()
 
 
 def ntriples_unquote_validate(input: str) -> str:
@@ -92,7 +98,7 @@ unquoters: Dict[str, Callable[[str], str]] = {
 }
 
 
-def make_correctness_tests(
+def make_unquote_correctness_tests(
     selectors: Iterable[str],
 ) -> Iterable[Tuple[str, str, str]]:
     """
@@ -100,12 +106,12 @@ def make_correctness_tests(
     `CORRECTNESS_PAIRS` that is suitable for use as pytest parameters.
     """
     for selector in selectors:
-        for quoted, unquoted in CORRECTNESS_PAIRS:
+        for quoted, unquoted in UNQUOTE_CORRECTNESS_PAIRS:
             yield selector, quoted, unquoted
 
 
 @pytest.mark.parametrize(
-    "unquoter_key, quoted, unquoted", make_correctness_tests(unquoters.keys())
+    "unquoter_key, quoted, unquoted", make_unquote_correctness_tests(unquoters.keys())
 )
 def test_unquote_correctness(
     unquoter_key: str,
@@ -124,7 +130,7 @@ QUAD_FORMATS = {"nquads"}
 
 @pytest.mark.parametrize(
     "format, quoted, unquoted",
-    make_correctness_tests(["turtle", "ntriples", "nquads"]),
+    make_unquote_correctness_tests(["turtle", "ntriples", "nquads"]),
 )
 def test_parse_correctness(
     format: str,
@@ -146,3 +152,67 @@ def test_parse_correctness(
     assert isinstance(obj, Literal)
     assert isinstance(obj.value, str)
     assert obj.value == unquoted
+
+
+EGNS = Namespace("http://example.com/")
+
+
+@pytest.mark.parametrize(
+    "format, char, escaped",
+    [
+        (format, char, escaped)
+        for format, (char, escaped) in itertools.product(
+            ["turtle"],
+            [
+                (r"x", r"x"),
+                (r"(", r"\("),
+                (r")", r"\)"),
+            ],
+        )
+    ],
+)
+def test_pname_escaping(format: str, char: str, escaped: str) -> None:
+    graph = Graph()
+    triple = (
+        URIRef(EGNS["prefix/John_Doe"]),
+        URIRef(EGNS[f"prefix/prop{char}"]),
+        Literal("foo", lang="en"),
+    )
+    graph.bind("egns", EGNS["prefix/"])
+    graph.add(triple)
+    data = graph.serialize(format=format)
+    pattern = re.compile(f"\\segns:prop{re.escape(escaped)}\\s")
+    logging.debug(
+        "format = %s, char = %r, escaped = %r, pattern = %r, data = %s",
+        format,
+        char,
+        escaped,
+        pattern,
+        data,
+    )
+    assert re.search(pattern, data) is not None
+
+
+# https://www.w3.org/TR/turtle/#grammar-production-PN_LOCAL_ESC
+# Not including %, as % should be used for percent encoding.
+PN_LOCAL_ESC_CHARS = r"_~.-!$&'()*+,;=/?#@"
+
+
+@pytest.mark.parametrize(
+    "format, char",
+    itertools.product(["turtle", "ntriples"], "A2c" + PN_LOCAL_ESC_CHARS),
+)
+def test_serialize_roundtrip(format: str, char: str) -> None:
+    graph = Graph()
+    triple = (
+        URIRef(EGNS["prefix/John_Doe"]),
+        URIRef(EGNS[f"prefix/prop{char}"]),
+        Literal("foo", lang="en"),
+    )
+    graph.add(triple)
+    graph.bind("egns", EGNS["prefix/"])
+    data = graph.serialize(format=format)
+    logging.debug("format = %s, char = %s, data = %s", format, char, data)
+    parsed_graph = Graph()
+    parsed_graph.parse(data=data, format=format)
+    GraphHelper.assert_sets_equals(graph, parsed_graph)

